scipy/special/_basic.py

```diff
@@ -3087,6 +3090,139 @@ def factorialk(n, k, exact=True):
+
+    References
+    ----------
+    .. [1] R. L. Graham, D. E. Knuth and O. Patashnik, "Concrete
+        Mathematics: A Foundation for Computer Science," Addison-Wesley
+        Publishing Company, Boston, 1989. Chapter 6, page 258.
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from scipy.special import stirling2
+    >>> k = np.array([3, -1, 3])
+    >>> n = np.array([10, 10, 9])
+    >>> stirling2(n, k)
+    array([9330, 0, 3025], dtype=object)
+
+    """
+    if exact:
+        # make a min-heap of unique (n,k) pairs
+        N, K = asarray(N), asarray(K)
+        if not (
+                np.issubdtype(N.dtype, np.integer)
+                or all(
+                    isinstance(n.take(0), int)
+                    for n in np.nditer(N, ['refs_ok'])
+                )
+        ):
+            raise TypeError("Argument `N` must contain only integers")
+        if not (
+                np.issubdtype(K.dtype, np.integer)
+                or all(
+                    isinstance(k.take(0), int)
+                    for k in np.nditer(K, ['refs_ok'])
+                )
+        ):
+            raise TypeError("Argument `K` must contain only integers")
+        nk_pairs = list(
+            set([(n.take(0), k.take(0))
+                 for n, k in np.nditer([N, K], ['refs_ok'])])
+        )
+        heapify(nk_pairs)
+        # base mapping for small values
+        snsk_vals = defaultdict(int)
+        for pair in [(0, 0), (1, 1), (2, 1), (2, 2)]:
+            snsk_vals[pair] = 1
+        n_old, n_row = 2, [0, 1, 1]
+        # for each pair in the min-heap, calculate the value, store for later
+        while nk_pairs:
+            n, k = heappop(nk_pairs)
+            if n < 2 or k > n or k <= 0:
+                continue
+            elif k == n or k == 1:
+                snsk_vals[(n, k)] = 1
+                continue
+            elif n != n_old:
+                num_iters = n - n_old
+                while num_iters > 0:
+                    n_row.append(1)
+                    # traverse from back to remove second row
+                    for j in range(len(n_row)-2, 1, -1):
+                        n_row[j] = n_row[j]*j + n_row[j-1]
+                    num_iters -= 1
+                snsk_vals[(n, k)] = n_row[k]
+            else:
+                snsk_vals[(n, k)] = n_row[k]
+            n_old, n_row = n, n_row
+        # for each pair in the map, fetch the value, and populate the array
+        it = np.nditer(
+            [N, K, None],
+            ['buffered', 'refs_ok'],
+            [['readonly'], ['readonly'], ['writeonly', 'allocate']],
+            op_dtypes=[object, object, object],
+        )
+        with it:
+            while not it.finished:
+                it[2] = snsk_vals[(int(it[0]), int(it[1]))]
+                it.iternext()
+            output = it.operands[2]
+            # If output is 0d array, convert to scalar.
+            if not output.ndim:
+                output = output.take(0)
```

Comment:
```suggestion
```
deleting this should make the function return an array in the 0d case

Reply:
If the input is scalar the output should be as well. I think just deleting this will cause 0d output for scalar input. You’ll need to check the input type.

Reply:
> If the input is scalar the output should be as well. I think just deleting this will cause 0d output for scalar input. You’ll need to check the input type.

yeah, I'll add the `output_is_scalar` change your suggesting.